rove_connector/src/lib.rs

```diff
@@ -0,0 +1,322 @@
+    }
+
+    out
+}
+
+impl Connector {
+    async fn fetch_one(
+        &self,
+        ts_id: i32,
+        time_spec: &TimeSpec,
+        num_leading_points: u8,
+        num_trailing_points: u8,
+    ) -> Result<DataCache, data_switch::Error> {
+        // TODO: matching intervals like this is a hack, but currently necessary to avoid
+        // SQL injection. Ideally we could pass an interval type as a query param, which would
+        // also save us the query_string allocation, but no ToSql implementations for intervals
+        // currently exist in tokio_postgres, so we need to implement it ourselves.
+        let interval = match time_spec.time_resolution {
+            x if x == RelativeDuration::minutes(1) => "1 minute",
+            x if x == RelativeDuration::hours(1) => "1 hour",
+            x if x == RelativeDuration::days(1) => "1 day",
+            _ => {
+                return Err(data_switch::Error::Other(Box::new(
+                    Error::UnhandledTimeResolution(time_spec.time_resolution),
+                )))
+            }
+        };
+
+        let (start_time, end_time) =
+            extract_time_spec(time_spec, num_leading_points, num_trailing_points)?;
+
+        // TODO: should this contain an ORDER BY? Actually I think it's not necessary since the
+        // order is dictated by the generated sequence
+        // TODO: should we drop ts_rule.timestamp from the SELECT? we don't seem to use it
+        // TODO: should we make this like the fetch_all query and regularize outside the query?
+        // I think this query might perform badly because the join against the generated series
+        // doesn't use the index optimally. Doing this would also save us the "interval" mess
+        let query_string = format!("SELECT data.obsvalue, ts_rule.timestamp \
+                FROM (SELECT data.obsvalue, data.obstime FROM data WHERE data.timeseries = $1) as data 
+                    RIGHT JOIN generate_series($2::timestamptz, $3::timestamptz, interval '{}') AS ts_rule(timestamp) \
+                        ON data.obstime = ts_rule.timestamp", interval);
+
+        let conn = self
+            .pool
+            .get()
+            .await
+            .map_err(|e| data_switch::Error::Other(Box::new(e)))?;
+
+        let data_results = conn
+            .query(query_string.as_str(), &[&ts_id, &start_time, &end_time])
+            .await
+            .map_err(|e| data_switch::Error::Other(Box::new(e)))?;
+
+        let cache = {
+            let mut values = Vec::with_capacity(data_results.len());
+
+            for row in data_results {
+                values.push(row.get(0));
+            }
+
+            DataCache::new(
+                vec![Timeseries {
+                    tag: ts_id.to_string(),
+                    values,
+                }],
+                // TODO: we need to either query to get the lat, lon, elev, or change olympian to
+                // accept not having them
```

Comment:
For a single timeseries it doesn't really make sense, right? Unless we have single checks that use the location in some way

Reply:
Yeah. In theory it could be used by climate_range_check, but I think even in that case PiM computes the ranges per station_id. I made an issue in olympian to track this